src/core/mac/sub_mac.cpp

```diff
@@ -585,7 +586,11 @@ void SubMac::HandleTransmitDone(TxFrame &aFrame, RxFrame *aAckFrame, Error aErro
     SetState(kStateReceive);
 
 #if OPENTHREAD_RADIO
-    if (aFrame.GetChannel() != aFrame.GetRxChannelAfterTxDone())
+    if (!mRxOnWhenIdle)
+    {
+        IgnoreError(Sleep());
```

Comment:
We cannot go to sleep in all cases; for example, if the tx frame is a data poll and the received acknowledgment indicates that a frame is pending.

The docs mention this behavior under `OT_RADIO_CAPS_RX_ON_WHEN_IDLE`: https://github.com/openthread/openthread/blob/0066d565afb6052c4a93eb32fc917cbcbb95cf48/include/openthread/platform/radio.h#L623-L631

If your goal is to implement `OT_RADIO_CAPS_RX_ON_WHEN_IDLE` functionality in RCP, it may require more code and some bigger changes. `SubMac` would practically need to implement a subset of the "operation" state tracking that `Mac` does.

My suggestion when `RxOnWhenIdle` on the radio was originally added in #9554 was to define a new `Mac`-level API for interaction with a radio chip (https://github.com/openthread/openthread/pull/9554#issuecomment-1775907319).

Reply:
Good point for `Data Request`! We may either add a new spinel property to specify the `DATA_POLL_TIMEOUT` for data request only or adding a field in `TxInfo` to specify how long the receiver should be on after TX done for all frame types. The latter looks more extensible. Given once the RCP firmware is released in a product, it's not as easy as the host to add new features. I'd prefer the latter approach or both. What do you think?

Reply:
Yakun, I think you added support for the tracking data polls in `SubMac`, but that will only partially support `OT_RADIO_CAPS_RX_ON_WHEN_IDLE`.

If the goal is to enable `OT_RADIO_CAPS_RX_ON_WHEN_IDLE` (in `SubMac` for RCP), it may be worth creating a separate pull request. Other things need to be implemented to enable this. We need to implement all the behaviors documented here:
https://github.com/openthread/openthread/blob/0066d565afb6052c4a93eb32fc917cbcbb95cf48/include/openthread/platform/radio.h#L614-L647

----


Generally, I am not fond of the `OT_RADIO_CAPS_RX_ON_WHEN_IDLE` feature. It was added mainly to support the behavior of certain SoC radios (due to their architecture).

We should be careful about enabling/using this feature for RCP. It may lead to unexpected/untested behaviors. As an example, I can refer to this issue https://github.com/openthread/openthread/issues/10302 (which is now addressed). 

I feel that many MAC functions are gradually being shifted to the radio, each adding more complicated `otPlatRadio` APIs and expected behaviors. I always prefer the radio APIs to be simple.

In my view, the better model is to define a new set of (async) MAC APIs for interaction with a new type of RCP which runs mac (maybe calling it MCP?). This will certainly be more work, but I think it is the right way to approach this.

Thoughts?

Reply:
I agree with you that there's a lot of changes need to fully implement rx-on-when-idle in the sub_mac. I've reduced the changes in the PR, so that it only avoid issuing `Receive()` while rx-on-when-idle is false for switching to RxChannelAfterTxDone.

Reply:
@bukepo, I see you reduced the scope of this change. Thanks. Does this enable an important function? I'm concerned that adding a small piece of the `mRxOnWhenIdle` functionality in RCP could cause confusion. I'm okay with adding it if it it critical and addresses something immediately, otherwise, I would recommend not adding it?

Reply:
The major motivation is to avoid putting the device into receive mode unintentionally. It's not a critical issue, but from previous experience, it would become more and more challenging to get coprocessor side issues fixed over time, so I'd prefer fixing found issues ASAP. Thoughts?